### ISSUE TITLE
multi grids is not working as expected

### ISSUE THREAD

**Issue by aliabdolali** [NOAA-EMC/WW3#417]
The _GFSv16_ wave setup has 3 grids (_gnh_10m, gsh_15m and aoc_9km_) with wind, ice and current forcings.
The forcings are:

1. WIND (hourly- plotted every 3 hrs)
![Forcing_WIND](https://user-images.githubusercontent.com/37336972/122323099-8c880a80-cef4-11eb-982a-ea5d332895c8.gif)
2. Current (hourly- plotted every 3 hrs)
![Forcing_Current](https://user-images.githubusercontent.com/37336972/122323096-8bef7400-cef4-11eb-89c9-2f3f86f0c334.gif)
3. ICE (daily)
![Forcing_ICE](https://user-images.githubusercontent.com/37336972/122323098-8c880a80-cef4-11eb-9b26-034ca84691c1.gif)

It was working fine prior to [e756361](https://github.com/NOAA-EMC/WW3/commit/e756361cee6c0e3bfac064376fcec3eba986719d) as shown in the following animation
**Expected behaviour (gfsv16 wave setup):**
![GFSv16_correct](https://user-images.githubusercontent.com/37336972/122323122-97db3600-cef4-11eb-92a6-4f720563ffad.gif)

However, it shows a weird significant wave height field if the top of NOAA/[develop](https://github.com/NOAA-EMC/WW3/commit/6cfe1bd0961354c778947da6b6a30f8711b54024) is used. The wave height drops significantly every few time steps.
**Strange HS (gfsv16 wave setup):**
![Multi_GFSv16](https://user-images.githubusercontent.com/37336972/122323201-b8a38b80-cef4-11eb-8b83-0b43dce61afe.gif)

I did some additional tests to make sure it is related to multiple grids. 
I ran a global grid with 1deg resolution using _ww3_multi_ and it worked as expected.

**Expected behaviour (single grid with 1 deg resolution):**
![Global1Deg](https://user-images.githubusercontent.com/37336972/122323273-d670f080-cef4-11eb-8403-02266e60a5f7.gif)

**To Reproduce**
There is a [Feature branch](https://github.com/aliabdolali/WW3/tree/fb_ufs_regtests) with _ww3_ufs1.1_ (1deg global grid) and _ww3_ufs1.2_ (GFSv16 multi grids) tests. The inputs are on ftp server, so if this feature branch is cloned, the `model/bin/ww3_from_ftp.sh` can be used to retrieve the inputs.

To reproduce them, the _regtest/bin/run_test_ can be used:
**ww3_ufs1.1 (working with [e756361](https://github.com/NOAA-EMC/WW3/commit/e756361cee6c0e3bfac064376fcec3eba986719d)  and top of noaa/[develop](https://github.com/NOAA-EMC/WW3/commit/6cfe1bd0961354c778947da6b6a30f8711b54024))**
`./bin/run_test -b slurm -c hera.intel -S -T -w work_a -m grdset_a -f -p mpirun -n 140 -t 4  -o all ../model ww3_ufs1.1`
and 
**ww3_ufs1.2 (working with the commits before [e756361](https://github.com/NOAA-EMC/WW3/commit/e756361cee6c0e3bfac064376fcec3eba986719d) and not working with the top of noaa/[develop](https://github.com/NOAA-EMC/WW3/commit/6cfe1bd0961354c778947da6b6a30f8711b54024))**
`./bin/run_test -b slurm -c hera.intel -S -T -w work_ali -m grdset_a -f -p mpirun -n 140  -t 4  -o all ../model ww3_ufs1.2`

**Note** 
For the commits prior to [482c500](https://github.com/NOAA-EMC/WW3/commit/482c500da51c2caf41ce3f8f011855bb4ab38d76), Flags for the two additional variables in ww3_multi.inp should be removed (Atmospheric momentum and Air density). 

Linked Pull Request: https://github.com/NOAA-EMC/WW3/pull/350

**Comment by JessicaMeixner-NOAA** [comment]
@aliabdolali how about the other multi-grid tests?   Are those working as expected? Something that is unique about this test I believe is some of the curvilinear grids maybe?

**Comment by aliabdolali** [comment]
> @aliabdolali how about the other multi-grid tests? Are those working as expected? Something that is unique about this test I believe is some of the curvilinear grids maybe?

I ran the multi setup with two rectilinear grids (gnh_10m and gsh_15m) excluding the curvilinear grid (aoc_9km) and it still shows strange behavior.

**Comment by JessicaMeixner-NOAA** [comment]
@aliabdolali  I went back and have looked more at https://github.com/NOAA-EMC/WW3/pull/264 and if you look at the matrix output I found: 
```
* test case: mww3_test_03; test run: ./work_PR2_UQ_MPI_e
*********************************************************
  found 51 files in base directory
  found 51 files in compare directory
     ww3_strt_low0.out are identical
     log.hgh0 are identical (filtered)
     out_grd.low0 differ (binary)
     out_pnt.points are identical (binary)
     restart.low0 differ (binary)
     ww3_grid_points.out are identical
     ww3_outp_tab50_points.out are identical
     restart001.hgh0 differ (binary)
     skipped time_count.txt
     ww3_outf_low0/ww3.68060608.hs differ. 
     ww3_outf_low0/ww3.68060603.hs differ. 
     ww3_outf_low0/ww3.68060610.hs differ. 
     ww3_outf_low0/ww3.68060609.hs differ. 
     ww3_outf_low0/ww3.68060605.hs differ. 
     ww3_outf_low0/ww3.68060602.hs differ. 
     ww3_outf_low0/ww3.68060606.hs differ. 
     ww3_outf_low0/ww3.68060611.hs differ. 
     ww3_outf_low0/ww3.68060600.hs are identical
     ww3_outf_low0/ww3.68060612.hs differ. 
     ww3_outf_low0/ww3.68060607.hs differ. 
     ww3_outf_low0/ww3.68060604.hs differ. 
     ww3_outf_low0/ww3.68060601.hs differ. 
     skipped finished

* test case: mww3_test_03; test run: ./work_PR1_MPI_e
*********************************************************
  found 51 files in base directory
  found 51 files in compare directory
     ww3_strt_low0.out are identical
     log.hgh0 are identical (filtered)
     out_grd.low0 differ (binary)
     out_pnt.points are identical (binary)
     restart.low0 differ (binary)
     ww3_grid_points.out are identical
     ww3_outp_tab50_points.out are identical
     restart001.hgh0 differ (binary)
     skipped time_count.txt
     ww3_outf_low0/ww3.68060608.hs differ. 
     ww3_outf_low0/ww3.68060603.hs differ. 
     ww3_outf_low0/ww3.68060610.hs differ. 
     ww3_outf_low0/ww3.68060609.hs differ. 
     ww3_outf_low0/ww3.68060605.hs differ. 
     ww3_outf_low0/ww3.68060602.hs differ. 
     ww3_outf_low0/ww3.68060606.hs differ. 
     ww3_outf_low0/ww3.68060611.hs differ. 

```
Both of which are not normal mww3_test_03 tests that normally have differences and is not the restart file where we know we have differences.  Perhaps there are other test cases which have this same problem? I haven't looked at output yet, but that's my next step.

**Comment by JessicaMeixner-NOAA** [comment]
Okay, so I ran the following: 
./bin/run_test -b slurm -c hera -S -T -s PR1_MPI -w work_PR1_MPI_e_both02 -o both  -m grdset_e  -f -p mpirun -n 24 ../model mww3_test_03
which for others is probably just simply: 
./bin/run_test  -s PR1_MPI -w work_PR1_MPI_e_both02 -o both  -m grdset_e  -f -p mpirun -n 24 ../model mww3_test_03

with commits: 

commit 9735c7de7928e592bbead898f7ebe2f3b01c40d3 (HEAD)
Author: Mickael Accensi <49198861+mickaelaccensi@users.noreply.github.com>
Date:   Fri Oct 23 16:40:22 2020 +0200

and 

commit e756361cee6c0e3bfac064376fcec3eba986719d (HEAD)
Author: Chris Bunney <48915820+ukmo-ccbunney@users.noreply.github.com>
Date:   Thu Nov 5 16:08:39 2020 +0000

and when I make a diff of the netcdf files I got this: 

![Screen Shot 2021-06-22 at 11 49 59 AM](https://user-images.githubusercontent.com/22549231/122957644-2239f480-d350-11eb-9f65-28ff0e7fed0a.png)

I've looked at the code in  PR#264 and nothing is jumping out at me as the cause of this, but it does appear to come into that commit just like the case @aliabdolali described above.

**Comment by ukmo-ccbunney** [comment]
@JessicaMeixner-NOAA  Thanks for spotting this. I will investigate. We're having HPC issues today, so it might be next week.

**Comment by ukmo-juan-castillo** [comment]
I am having a look at this too.

**Comment by ukmo-juan-castillo** [comment]
I think I have found the origin of the problem. With the latest changes, the format of the restart file changes, so the old restart file present in the restes is not being read properly. The solution would be generating an equivalent restart file valid for the new code.

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-juan-castillo thanks for looking into this!   Where are you seeing the old restart files from the regtests?  I don't see any binary files in the regression tests except for ww3_tp2.12 partition.ww3 files which are not in these tests.   I did think it was perhaps an issue with the restart file changes, but I was thinking maybe something got reset to zero when writing the restart files unintentionally somehow, but hadn't followed up on that theory yet.

**Comment by ukmo-juan-castillo** [comment]
The issue is definitely in the code reading/writing the restart file, but I did not investigate in depth yet. Hopefully I will give you more details soon (I am still testing), I just told you to revisit the restart file in case that was a quick solution, but I did not check if there was a restart file in the test (I assumed there was), nor checked its format.

**Comment by JessicaMeixner-NOAA** [comment]
Sounds good @ukmo-juan-castillo I  really appreciate your help and the suggestions.  Let me know if there's anything I can do to help look at things or for extra testing, etc.

**Comment by aliabdolali** [comment]
@JessicaMeixner-NOAA and @ukmo-juan-castillo Good catch,
I reran the ww3_ufs1.2 test and change ww3_multi with no restart output, then hs field looked fine. It seems it reinitiates at each time the model writes a restart file.
This bug just affects the ww3_multi cases with more than 1 grid (ww3_ufs1.1 with 1 single grid works fine with restart output).

**Comment by ukmo-juan-castillo** [comment]
I think I have found the problem. It has to do with the IF conditions in section 2.b of the subroutine W3MPIO. I will run some test to make sure, and if the fix works I will create a branch and a pull request in the NOAA repository.

**Comment by ukmo-juan-castillo** [comment]
It looks like the fix is working. I will create a branch from the latest version of develop, and run all the regtests. I will also run a well known coupled configuration of the Met Office (the changes may affect coupled configurations). Once the changes are implemented in the branch, please test it to make sure the differences you observe disappear.

**Comment by ukmo-juan-castillo** [comment]
I think I know what happens now. The fix I propose seems right, and it will affect the multi tests writing a restart file, which also explains the observed changes in ww3_tp2.17. When applying the fix to version e756361cee6c0e3bfac064376fcec3eba986719d there are no changes in the mww3_test_04 test though. This happens because there was a bug in the code affecting these tests that was only fixed in version 6d95c334eebdb55d3ac1211240fa24d9f959cc83, in particular changes in the routine W3SETO to assign the pointers IRQRSS and VAAUX for all processors and not just when NRQRS==0:

https://github.com/NOAA-EMC/WW3/commit/6d95c334eebdb55d3ac1211240fa24d9f959cc83#diff-939f91a8acb5a641add991630802f1bfd7f2d322046224043be109fd2a15ae91

Only after this bug is fixed the current bug fix will change results in mww3_test_04.
